internal/xds/translator/ratelimit.go

```diff
@@ -495,31 +521,103 @@ func addRateLimitDescriptor(
 	serviceDescriptors []*rlsconfv3.RateLimitDescriptor,
 	domain string,
 	domainDescriptors map[string][]*rlsconfv3.RateLimitDescriptor,
+	includeShared bool,
 ) {
-	var key, value string
+	if !isValidGlobalRateLimit(route) || len(serviceDescriptors) == 0 {
+		return
+	}
 
-	if isSharedRateLimit(route) {
-		// For shared rate limits, use traffic policy name key/value
-		key = route.Traffic.Name
-		value = route.Traffic.Name
-	} else {
-		// For non-shared rate limits, use route descriptor key/value
-		key = getRouteDescriptor(route.Name)
-		value = getRouteDescriptor(route.Name)
+	getParent := func(key string) *rlsconfv3.RateLimitDescriptor {
+		for _, d := range domainDescriptors[domain] {
+			if d.Key == key {
+				return d
+			}
+		}
+		p := &rlsconfv3.RateLimitDescriptor{Key: key, Value: key}
+		domainDescriptors[domain] = append(domainDescriptors[domain], p)
+		return p
 	}
 
-	descriptor := &rlsconfv3.RateLimitDescriptor{
-		Key:         key,
-		Value:       value,
-		Descriptors: serviceDescriptors,
+	for i, rule := range route.Traffic.RateLimit.Global.Rules {
+		if i >= len(serviceDescriptors) || (includeShared != isRuleShared(rule)) {
```

Comment:
curious why the logic around descriptor creation has increased, shouldnt it be the same and leverage the rule name for descriptor key and value ?